proposals/003-authentication-api.md

```diff
@@ -0,0 +1,650 @@
+
+## Current situation
+
+The `Filter` and `FilterContext` APIs currently don't directly expose any authenticated client identity information.
+Specifically:
+
+* If proxy uses client mTLS, then filters don't have access to a `Subject` or `Principal` corresponding to the client's TLS client certificate.
+* If clients are authenticating using SASL, the only way a `Filter` can know about that is by intercepting those frames.
+    - identity-using filters in the chain must _each_ implement SASL passthrough inspection. 
+    - but this is usually incompatible with use of a filter performing SASL Termination or SASL Initiation.
+
+## Motivation
+
+The lack of API support makes implementing client identity aware plugins difficult, or impossible.
+
+Goals: 
+
+* Allow the possibility for new kinds of KRPC intercepting plugins in the future by not assuming that `Filters` are the only kind of KRPC intercepting plugin. We'll use the term **plugin**, unless saying something specifically about `Filters`.
+* Enable plugins to access a client's identity using a single, consistent API, irrespective of which authentication mechanism(s) are being used, TLS or SASL, and whether they're implemented by the proxy runtime (in the TLS case), or a prior plugin in the chain (in the SASL termination case).
+* Allow access to TLS- or SASL-specific details by plugins should they need them.
+* Don't require a plugin to handle `SaslAuthenticate` unless it is performing SASL termination or initiation.
+* Provide a flexible API to make serving niche use cases possible (though perhaps not simple).
+* Drop support for the "raw" (i.e. not encapsulated within the Kafka protocol) support for SASL, as [Kafka itself has does from Kafka 4.0](https://cwiki.apache.org/confluence/display/KAFKA/KIP-896%3A+Remove+old+client+protocol+API+versions+in+Kafka+4.0)
+
+## Proposal
+
+### API for Filters to access client TLS information
+
+TLS (in contrast to SASL) is handled entirely by the proxy runtime. 
+By the time a `Filter` is instantiated the proxy already has an established TLS connection.
+All that's required is an API for exposing appropriate details to `Filters`.
+
+The following method will be added to the existing `FilterContext` interface:
+
+```java
+    /**
+     * @return The TLS context for the client connection, or empty if the client connection is not TLS.
+     */
+    Optional<ClientTlsContext> clientTlsContext();
+```
+
+Where `ClientTlsContext` is a new interface in the new package `io.kroxylicious.proxy.tls`:
+
+```
+package io.kroxylicious.proxy.tls;
+
+import java.security.cert.X509Certificate;
+import java.util.Optional;
+
+public interface ClientTlsContext {
+    /**
+     * @return The TLS server certificate that the proxy presented to the client during TLS handshake.
+     */
+    X509Certificate proxyServerCertificate();
```

Comment:
Is a single certificate sufficient for the use-cases?  What if the server certificate is using intermediates?  I wonder if returning an ordered list containing the whole chain would be a better API.  There would be a guarantee that the chain would always have at least one certification (the self signed case).

Reply:
Ok, we can change the API later if the need arises.